src/backend/distributed/commands/alter_table.c

```diff
@@ -1122,10 +1125,20 @@ ReplaceTable(Oid sourceId, Oid targetId, List *justBeforeDropCommands,
 									quote_qualified_identifier(schemaName, sourceName))));
 		}
 
-		appendStringInfo(query, "INSERT INTO %s SELECT * FROM %s",
-						 quote_qualified_identifier(schemaName, targetName),
-						 quote_qualified_identifier(schemaName, sourceName));
-		ExecuteQueryViaSPI(query->data, SPI_OK_INSERT);
+		/*
+		 * Since we wrap target columns with parenthesis, skip executing
+		 * INSERT .. SELECT if relation has no columns that require moving
+		 * data. Otherwise, postgres could error out.
+		 */
+		if (AnyColumnsRequireDataCopy(sourceId))
```

Comment:
do we have a test where `AnyColumnsRequireDataCopy ` returns false?

Reply:
not added in this pr but yes
https://github.com/citusdata/citus/blob/cacb76d2c6a70383ad5d7c597daa2298e446ac2b/src/test/regress/expected/fkeys_between_local_ref.out#L117..L118

but I can add such a test in this pr too

Reply:
Yes, lets have one. Thanks!

Reply:
a08014663729b3f6b9788c26ed92a3330f721b6b

Reply:
Ok, it seems that this logic is broken:
```SQL
CREATE TABLE t1 (a int, b int generated always as (a + 1) stored);

INSERT INTO t1 (a) VALUES (1), (2);


ALTER TABLE t1 DROP COLUMN a;

SELECT * FROM t1;
--
(2 rows)

Time: 0.469 ms

 select alter_table_set_access_method('t1', 'columnar');

-- !!---!! This should have returned two rows
SELECT * FROM t1;
--
(0 rows)

Time: 0.935 ms

```

Reply:
tbh, first one sounds like a postgres bug to me at first glance, but in the end, it should be normal to see `0` rows after altering the table since we create the table from scratch ?

Reply:
seems that below was crashing on master, but not on this branch since we skip `insert select`:
```sql
create table local(y int);
alter table local drop column y;
select alter_table_set_access_method('local', 'columnar');
```

Reply:
> moreover, seems that below was crashing on master, but not on this branch since we skip insert select:

Good that we catch this

> tbh, first one sounds like a postgres bug to me at first glance (or I'm missing something, is it the expected behavior ?)

Why is so? The table has two rows and PG gives that information to us? This behavior is valid regardless of generated columns. If you drop all columns of a table, you'd see the row count is unchanged.

I guess if the table has no columns, we should do?
```SQL
insert into t2 select * from t1;
```

Reply:
to handle https://github.com/citusdata/citus/pull/4616#discussion_r568646297:

when no columns available, we should execute plain `INSERT INTO temp_table SELECT * FROM input_table` so that we don't break pg output but not sure if it's worth ?

Reply:
Yes, I think it is definitely worth, we cannot return wrong results. We need good comments around it. And, maybe logic should be changed to:

```C
if (HasAnyGeneratedStoredColumns())
{
     build list, do insert select with the list
}
else
{
  do the regular thing
}
```

We should add a test for this scenario

Reply:
0fc3153e1f7a1cb8c2a084e728dbfb5995516109